### PR TITLE
feat(lib/runtime): remove runtime type reflection in host module function definitions (with Generics)

### PR DESCRIPTION
**Pull request by ec2**
## Changes
Same as https://github.com/ChainSafe/gossamer/pull/3960, but it uses generics instead to enforce type safely. Keeping this as draft because I havent tested the impact on performance of using generics

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

### CODE REVIEW

**Review by edwardmack — APPROVED** [review]
This looks great.  You mention potential performance differences between this and #3960.  Are you planning to get some metrics to determine if #3962 or #3960 should be merged?

**Comment by ec2** [comment]
> This looks great. You mention potential performance differences between this and #3960. Are you planning to get some metrics to determine if #3962 or #3960 should be merged?

Yeah. @EclesioMeloJunior has a script here https://github.com/ChainSafe/gossamer/tree/development/scripts/retrieve_block

I will try it out and see which is better probably by today or Monday. Then I will post results and let y'all decide which to merge and which to close. This one is better for maintainability and readability.

**Comment by ec2** [comment]
Or you can merge #3960 first because its for sure faster. And then this one can be merged based on measured performance. In that way, we can have both in the commit history.

**Comment by EclesioMeloJunior** [comment]
@ec2 since the https://github.com/ChainSafe/gossamer/pull/3960 was already merged, can we merge this one?

File: lib/runtime/wazero/imports.go [hunk 1]
```diff
@@ -116,6 +116,88 @@ func ext_crypto_ecdsa_generate_version_1(ctx context.Context, m api.Module, _ ui
...
+	uint32 | uint64 | int32
+}
+
+func wrap0(f func(ctx context.Context, m api.Module)) api.GoModuleFunc {
+	return func(ctx context.Context, m api.Module, stack []uint64) {
+		f(ctx, m)
+	}
+}
+
+func wrap1[T0 WasmParamType](f func(ctx context.Context, m api.Module, a T0)) api.GoModuleFunc {
```
Comment (EclesioMeloJunior) [inline lib/runtime/wazero/imports.go]: You can define a type for the `f` function like to make the functions less verbose and the types clear in the top of the file. 

```go
type noArg func(ctx context.Context, m api.Module)
type singleArg[T WasmParamType] func(ctx context.Context, m api.Module, arg1 T)
...
...
type noArgWithRet[R WasmParamType] func(ctx context.Context, m api.Module) R
...

func wrap0(f noArg) api.GoModuleFunc { ... }

func wrap1[T0 WasmParamType](f singleArg[T]) api.GoModuleFunc { ... }

func wrap0ret[R WasmParamType](f noArgWithRet[R]) api.GoModuleFunc {
```

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]
LGTM, really great changes, makes the code a lot more type safe, thanks!

**Comment by EclesioMeloJunior** [comment]
Depends on https://github.com/ChainSafe/gossamer/pull/4041

**Review by jimjbrettj — APPROVED** [review]
nice work!

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**